### PR TITLE
[Error Handling] Introducing new error handling method - `Result`

### PR DESCRIPTION
**Pull request by AAsteria**
# Pull Request Template

## Description

**浅谈一下各种语言的Error Handling**:
- JavaScript的函数特性中, 在JavaScript中, **任何`Error`都可以在任何地方抛出，而不需要特殊的声明**. 意味着每个function都有可能报错.
```js
function mightThrow() {
        throw new Error('I told you this might throw');
}

try {
        mightThrow();
} catch (err) {
        console.log(err);
}
```
- C++中每个函数也都可能`throw Exception`, 但是可以通过关键字`noexcept`来标注该函数不会throw. **任何`Exception`都可以在任何地方抛出，而不需要特殊的声明**:
```cpp
// Deprecated in C++11 (好像是因为性能等诸多原因, 该规范被扔了)
void oldFunction() throw (std::runtime_error) {
    // ...
}

// C++11 and later
void newFunction() noexcept {
    // ...
}
```
- Java提供**声明所有的异常方案**:
```java
public void riskyMethod() throws IOException, SQLException {
    // some code that might throw IOException or SQLException
}

public static void main(String[] args) {
      try {
            riskyMethod();
      } catch (IOException e) {
            // ...
      } catch (SQLException e) {
            // ...
      }
}
```
- Go直接抛弃了try-catch和throw, 选择提供**多返回值方案**:
```go
func SomeFunction() (ResultType, error) {
    // ...
    if somethingWrong {
        return nil, errors.New("something went wrong")
    }
    return result, nil
}

func main() {
    result, err := SomeFunction()
    if err != nil {
        // Handle the error
        log.Fatal(err)
    }
    // Continue with result
}
```
- 而Rust则采用和Go类似的方案, 但是更进一步, 将Compiler和enum做结合, 提供了`Result`方案:
```rust
use std::fs;

fn read_file(filename: &str) -> Result<String, std::io::Error> {
    fs::read_to_string(filename)
}

match read_file("path_to_file.txt") {
    Ok(content) => println!("File content: {}", content),
    Err(error) => println!("Failed to read file: {}", error),
}
```

>  try-catch的痛处
> 1. 作为client (upper layer) 不知道某个函数会不会报错 -> 在调用的时候, client可能会忘记catch error / 忘记写处理error的代码.
> 2. Client能catch到的err不知道是什么类型 -> 在TypeScript中catch到的Error默认是`unknown`类型 -> 对于部分语言而言, 我们需要手动检查函数会报什么类型的Error/Eception.

**旧方案**
> 1. 写Documentation来标记该函数是否会throw. (BAD) Documentation可以out of update.
> 2. 我们手动检查catch到的Error是什么类型. (BAD) 这意味着需要更多的沟通成本, client可能不清楚函数的内部implementation.

**新方案**
引进Rust的`Result`方案.
> - 首先, 我们通过ESLint来禁止使用`throw new Error()`, 来强行将JavaScript的特性修正成: 每个function默认不会`throw`.

> 1. 引进新类型`type Result<T, E> = Ok<T> | Err<E>`来描述该函数的返回值. 意味着该函数可能会出问题.
> 2. 利用Flow-Based Type Narrowing + ESLint来强迫用户必须处理`Result<T, E>`, 消除了client可能忘记处理`Result<T, E>`的可能性.

**附录**
1. [为什么Go选择了一个正确的Exceptions的处理思路](https://dave.cheney.net/2012/01/18/why-go-gets-exceptions-right)
2. From the [HackerNews thread](https://news.ycombinator.com/item?id=26191006) relating to "Where Everything Went Wrong: Error Handling and Error Messages in Rust (2020)":
```
Error handling has been wrong since the beginning, and has continued to be wrong ever since.

First, we had error codes. Except these were wrong because people forget all the time to check them.

Then we had exceptions, which solved the problem of people forgetting to check by crashing the app.

Then the Java team got the bright idea to have checked exceptions, which at first helped to mitigate crashes
from uncaught exception, but caused an explosion in thrown exception signatures, culminating in "catch Throwable".

Back to square one.

Then we got multi-return error objects, maybes, panics, and all sorts of bright ideas that fail to understand the basic premises of errors:

  Any error system that relies upon developer discipline will fail because errors will be missed.

  Any error system that handles all errors the same way will fail because there are some errors we can ignore,
  and some errors we must not ignore. And what's ignorable/retriable to one project is not ignorable/retriable to another.

Attempting to get the complete set of error types that any given call may raise is a fool's errand because of the halting 
problem it eventually invokes. Forcing people to provide such a list results in the Java problem for the same reason.

It's a hard problem, which is why no one has solved it yet. 
```
 

**Issue(s) Addressed**:
- never-throw repo: https://github.com/supermacro/neverthrow#resultmatch-method

## Changes
**Type of Changes**: new feature / breaking change
1. [ ] Bringing Rust's `Result` solution to TypeScript. In other words, we will create our own `Result` enum (or types).
2. [ ] `AsyncResult` for asynchronous programming.
3. [ ] 利用ESLint来:
    1. [ ] 禁止使用`throw`和`try-catch`的使用. 强行修正JavaScript的语言特性.
    2. [ ] Forced to consume the `Result` in one of the following three ways: calling `.match`, `.unwrapOr`, and `.unwrap` (unsafe). 
    - > A common problem with using return values to indicate errors is that it is easy to ignore the return value, thus failing to handle the error. With the help of ESLint, which will cause the compiler to issue an error when a `Result` is ignored. This ensures that we are explicitly handling the error of the `Result`.

## Checklist:
- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

## Reviewers Checklist:
- [ ] Code follows project guidelines and clean code principles
- [ ] Tested on various resolutions and browsers
- [ ] Unit tests have been updated or added, if necessary
- [ ] Documentation has been updated

### CODE REVIEW

**Review by Bistard — COMMENTED** [review]
fixed MacOS platform check

**Comment by Bistard** [comment]
来自于`neverThrow`的一个[issue](https://github.com/supermacro/neverthrow/issues/444) - 讨论能否把`?`语法从Rust搬到TypeScript中.

我在下面给出了我的回答 - 可以带过来, 但是很丑, 不够elegant, 所以放弃了.

**Comment by Bistard** [comment]
来自于`neverThrow`的一个[issue](https://github.com/supermacro/neverthrow/issues/444) - 讨论能否把?语法从Rust搬到TypeScript中.

我在下面给出了我的回答 - 可以带过来, 但是很丑, 不够elegant, 所以放弃了.

File: src/base/common/error.ts
Comment (Bistard) [inline src/base/common/error.ts:1]: `Ok` and `Err` must **BOTH** require two generic parameters: `T` and `E` to have a better type narrowing ability.

**Review by JM-001 — COMMENTED** [review]
Carefully reviewed error.ts

**Pull request merged**